# Лабораторная работа 6. Метод квазиэквивалентного укрупнения состояний многомерных марковских процессов размножения-гибели
## Задание 1

$K$ программистов могут писать программы для выполнения на одном из $M$ серверов,
при этом программа не попадает сразу на сервер,
а обрабатывается на одном из $N$ специальных компьютеров,
которые проверяют отсутствие вирусов.
Интенсивность работы программистов $\lambda$,
интенсивность работы компьютеров-антивирусов $\nu$,
интенсивность работы основных серверов $\mu$,
программа оказывается с вирусом с вероятностью $p$.
Если программа с вирусом, она получает отказ обслуживания на основных
серверах. Для компьютеров по проверке вирусов имеется ограничение по длине
очереди $m_1$ , для основных серверов ограничение по длине очереди $m_2$.

- Нарисовать граф состояний системы, учитывая количество программистов,
которые пишут программу, количество программ на компьютерах-антивирусах,
количество программ на серверах;
- Написать уравнения Колмогорова для вероятностей состояний, финальных
вероятностей;
- Тремя способами (экспериментально, методом укрупнения состояний, по
уравнениям Колмогорова) найти основные характеристики эффективности СМО:
  - среднее время пребывания заявки в системе,
  - абсолютную пропускную способность,
  - среднее число заявок в системе.

In [2]:
Variant <- 5
set.seed(Variant)
K <- sample(c(3:6), 1)
M <- sample(c(1:3), 1)
N <- sample(c(1:3), 1)
lambda <- runif(1)
mu <- runif(1)
nu <- runif(1)
p <- runif(1)
m2 <- sample(c(0:2), 1)
m1 <- sample(c(0:2), 1)
View(data.frame(K, M, N, lambda, mu, nu, p, m1, m2))

K M N lambda    mu        nu        p       m1 m2
1 4 3 1 0.2843995 0.1046501 0.7010575 0.52796 1  0

Введем следующие состояния, приняв $S_{k,q,n,m}$ за состояние с $k$ активных
программистов, $q$ заявок в очереди, $n$ заявок на проверке на вирусы и $m$
заявок на основном сервере:
- $S_{4000}$ - все программисты пишут программу, сервера и компьютеры
свободны;
- $S_{3010}$ - три программиста пишут программу, одна программа проверяется на
вирус;
- $S_{3001}$ - три программиста пишут программу, одна программа выполняется
на сервере;
- $S_{2110}$ - два программиста пишут программу, одна программа проверяется на
вирус, одна стоит в очереди;
- $S_{2011}$ - два программиста пишут программу, одна программа проверяется на
вирус, одна выполняется на сервере;
- $S_{2002}$ - два программиста пишут программу, две выполняются на сервере;
- $S_{1111}$ - один программист пишет программу, одна программа стоит
в очереди, одна программа проверяется на вирус, две выполняются на сервере;
- $S_{1003}$ - один программист пишет программу, три выполняются на
- $S_{1012}$ - один программист пишет программу, одна программа проверяется
на вирус, две выполняются на сервере;
- $S_{1003}$ - один программист пишет программу, три выполняются на
сервере;
- $S_{0112}$ - все программисты ожидают ответа, одна программа стоит
в очереди, одна программа проверяется
на вирус, две выполняются на сервере;
- $S_{0013}$ - все программисты ожидают ответа, одна программа проверяется на
вирус, три выполняются на сервере;

Тогда граф состояний будет выглядеть:
![graph](./State_graph.png)

### Теоретически по уравнениям Колмогорова

По графу составим уравнения Колмогорова:
$$
\begin{cases}
\frac{dP_{4000}(t)}{dt} = -4 \lambda P_{4000}(t) + p \nu P_{3010}(t) + \mu P_{3001}(t) \\
\frac{dP_{3010}(t)}{dt} = -(3 \lambda + p\nu + (1-p)\nu) P_{3010}(t) + 4 \lambda P_{4000}(t) + p \nu P_{2110}(t) + \mu P_{2011}(t) \\
\frac{dP_{2110}(t)}{dt} = -(p\nu + (1-p)\nu) P_{2110}(t) + 3 \lambda P_{3010}(t) + \mu P_{1111} \\
\frac{dP_{3001}(t)}{dt} = -(3 \lambda + \mu) P_{3001}(t) + p \nu P_{2011} + 2 \mu P_{2002} + (1-p) \nu P_{3010}(t) \\
\frac{dP_{2011}(t)}{dt} = -(2 \lambda + p\nu + (1-p)\nu + \mu) P_{2011}(t) + 3 \lambda P_{3001}(t) + p \nu P_{1111}(t) + 2 \mu P_{1012}(t) + (1-p)\nu P_{2110}(t) \\
\frac{dP_{1111}(t)}{dt} = -(p\nu + (1-p)\nu + \mu) P_{1111}(t) + 2 \lambda P_{2011}(t) + 2 \mu P_{0112} \\
\frac{dP_{2002}(t)}{dt} = -(2 \lambda + 2 \mu) P_{2002}(t) + p \nu P_{1012} + 3 \mu P_{1003} + (1-p) \nu P_{2011}(t) \\
\frac{dP_{1012}(t)}{dt} = -(\lambda + p\nu + (1-p)\nu + 2 \mu) P_{1012}(t) + 2 \lambda P_{2002}(t) + p \nu P_{0112}(t) + 3 \mu P_{0013}(t) + (1-p)\nu P_{1111}(t) \\
\frac{dP_{0112}(t)}{dt} = -(p\nu + (1-p)\nu + 2 \mu) P_{0112}(t) + \lambda P_{1012}(t) \\
\frac{dP_{1003}(t)}{dt} = -(\lambda + 3 \mu) P_{1003}(t) + p\nu P_{0013}(t) + (1-p)\nu P_{1012}(t) \\
\frac{dP_{0013}(t)}{dt} = -(p\nu + 3 \mu) P_{0013}(t) + \lambda P_{1003}(t) + (1-p)\nu P_{0112}(t)
\end{cases}
$$

и уравнение нормировки:
$$
P_{4000}(t) + P_{3010}(t) + P_{2110}(t) + P_{3001}(t) + P_{2011}(t) + P_{1111}(t) + P_{2002}(t) + P_{1012}(t) + P_{0112}(t) + P_{1003}(t) + P_{0013}(t) = 1
$$

Решим эти уравнения с помощью пакета deSolve
(см. 75 стр. [документации](chrome-extension://gfbliohnnapiefjpjlpjnehglfpaknnc/pages/pdf_viewer.html?r=https://cran.r-project.org/web/packages/deSolve/deSolve.pdf)):

In [3]:
if (!require("deSolve")) {
    install.packages("deSolve")
}
library(deSolve)

ode_system_equations <- function(Time, State, Pars) {
    with(as.list(c(State, Pars)), {
        dP_4000 <- -4 * lambda * abs(P_4000) + p * nu * abs(P_3010) + mu * abs(P_3001)
        dP_3010 <- -(3 * lambda + p * nu + (1-p) * nu) * abs(P_3010) + 4 * lambda * abs(P_4000) + p * nu * abs(P_2110) + mu * abs(P_2011)
        dP_2110 <- -(p * nu + (1-p) * nu) * abs(P_2110) + 3 * lambda * abs(P_3010) + mu * abs(P_1111)
        dP_3001 <- -(3 * lambda + mu) * abs(P_3001) + p * nu * abs(P_2011) + 2 * mu * abs(P_2002) + (1-p) * nu * abs(P_3010)
        dP_2011 <- -(2 * lambda + p* nu + (1-p) * nu + mu) * abs(P_2011) + 3 * lambda * abs(P_3001) + p * nu * abs(P_1111) + 2 * mu * abs(P_1012) + (1-p) * nu * abs(P_2110)
        dP_1111 <- -(p * nu + (1-p) * nu + mu) * abs(P_1111) + 2 * lambda * abs(P_2011) + 2 * mu * abs(P_0112)
        dP_2002 <- -(2 * lambda + 2 * mu) * abs(P_2002) + p * nu * abs(P_1012) + 3 * mu * abs(P_1003) + (1-p) * nu * abs(P_2011)
        dP_1012 <- -(lambda + p * nu + (1-p) * nu + 2 * mu) * abs(P_1012) + 2 * lambda * abs(P_2002) + p * nu * abs(P_0112) + 3 * mu * abs(P_0013) + (1-p) * nu * abs(P_1111)
        dP_0112 <- -(p * nu + (1-p) * nu + 2 * mu) * abs(P_0112) + lambda * abs(P_1012)
        dP_1003 <- -(lambda + 3 * mu) * abs(P_1003) + p * nu * abs(P_0013) + (1-p) * nu * abs(P_1012)
        dP_0013 <- -(p * nu + 3 * mu) * abs(P_0013) + lambda * abs(P_1003) + (1-p) * nu * abs(P_0112)

        # Specifying list of derivatives.
        return(
            list(c(
                dP_4000,
                dP_3010,
                dP_2110,
                dP_3001,
                dP_2011,
                dP_1111,
                dP_2002,
                dP_1012,
                dP_0112,
                dP_1003,
                dP_0013
            ))
        )
    })
}

Loading required package: deSolve



In [4]:
pars <- NULL
yini <- c(
    P_4000 = 1,
    P_3010 = 0,
    P_2110 = 0,
    P_3001 = 0,
    P_2011 = 0,
    P_1111 = 0,
    P_2002 = 0,
    P_1012 = 0,
    P_0112 = 0,
    P_1003 = 0,
    P_0013 = 0
)


INITIAL_TIME <- 0
FINISH_TIME <- 10000

ACCURACY <- 0.1
times <- seq(INITIAL_TIME, FINISH_TIME, by = ACCURACY)
output <- ode(yini, times, ode_system_equations, pars)
output

time    P_4000     P_3010     P_2110      P_3001      P_2011      
 [1,] 0.0     1.0000000  0.00000000 0.000000000 0.000000000 0.000000e+00
 [2,] 0.1     0.8943329  0.09956341 0.004337917 0.001668987 9.464439e-05
 [3,] 0.2     0.8030898  0.17474431 0.015529917 0.005932274 6.724613e-04
 [4,] 0.3     0.7239824  0.23064326 0.031324322 0.011891584 2.015504e-03
 [5,] 0.4     0.6551201  0.27134327 0.049997983 0.018881856 4.248648e-03
 [6,] 0.5     0.5949373  0.30010787 0.070245868 0.026416920 7.390385e-03
 [7,] 0.6     0.5421356  0.31953997 0.091093152 0.034147280 1.139025e-02
 [8,] 0.7     0.4956323  0.33171972 0.111819712 0.041824995 1.615764e-02
 [9,] 0.8     0.4545265  0.33829512 0.131911174 0.049281876 2.157753e-02
[10,] 0.9     0.4180629  0.34057759 0.151007423 0.056406481 2.752732e-02
[11,] 1.0     0.3856079  0.33960458 0.168869702 0.063130560 3.388410e-02
[12,] 1.1     0.3566274  0.33619570 0.185351293 0.069416912 4.053129e-02
[13,] 1.2     0.3306702  0.33099610 0.200375380 0.075250780 4.736187e-02
[14,] 1.3     0.3073534  0.32451165 0.213917396 0.080633165 5.428044e-02
[15,] 1.4     0.2863511  0.31713717 0.225991239 0.085575877 6.120392e-02
[16,] 1.5     0.2673847  0.30917923 0.236638376 0.090097763 6.806153e-02
[17,] 1.6     0.2502152  0.30087425 0.245919503 0.094222004 7.479431e-02
[18,] 1.7     0.2346369  0.29240309 0.253908073 0.097974124 8.135422e-02
[19,] 1.8     0.2204717  0.28390272 0.260685341 0.101380569 8.770311e-02
[20,] 1.9     0.2075654  0.27547547 0.266336620 0.104467713 9.381165e-02
[21,] 2.0     0.1957836  0.26719650 0.270948478 0.107261187 9.965821e-02
[22,] 2.1     0.1850087  0.25911972 0.274606673 0.109785442 1.052278e-01
[23,] 2.2     0.1751378  0.25128248 0.277394670 0.112063484 1.105111e-01
[24,] 2.3     0.1660801  0.24370932 0.279392605 0.114116740 1.155035e-01
[25,] 2.4     0.1577556  0.23641494 0.280676597 0.115965004 1.202042e-01
[26,] 2.5     0.1500934  0.22940654 0.281318322 0.117626457 1.246158e-01
[27,] 2.6     0.1430306  0.22268565 0.281384796 0.119117714 1.287434e-01
[28,] 2.7     0.1365112  0.21624964 0.280938296 0.120453912 1.325940e-01
[29,] 2.8     0.1304851  0.21009279 0.280036404 0.121648800 1.361762e-01
[30,] 2.9     0.1249076  0.20420721 0.278732125 0.122714849 1.394997e-01
[31,] ⋮       ⋮          ⋮          ⋮           ⋮           ⋮           
[32,]  9997.1 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[33,]  9997.2 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[34,]  9997.3 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[35,]  9997.4 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[36,]  9997.5 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[37,]  9997.6 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[38,]  9997.7 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[39,]  9997.8 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[40,]  9997.9 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[41,]  9998.0 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[42,]  9998.1 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[43,]  9998.2 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[44,]  9998.3 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[45,]  9998.4 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[46,]  9998.5 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[47,]  9998.6 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[48,]  9998.7 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[49,]  9998.8 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[50,]  9998.9 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[51,]  9999.0 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[52,]  9999.1 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[53,]  9999.2 0.02198278 0.03951838 0.06299633  0.09919352  0.1251528   
[54,]  9999.3 0.02198278 0.03951838 0.06299633  0.09919352  0.12

При $t \rightarrow{} \infty$ (взяли последний результат, брать значения $t$
выше нет смысла, так как видно, что значения $\forall P_i$ в конце таблицы
сходятся):

In [5]:
results <- output[nrow(output), 2:12]
results

P_4000     P_3010     P_2110     P_3001     P_2011     P_1111     P_2002 
0.02198278 0.03951838 0.06299633 0.09919352 0.12515277 0.09982820 0.17014707 
    P_1012     P_0112     P_1003     P_0013 
0.14139934 0.04417373 0.12307252 0.07253537

Подтвердим, что $\sum P_i = 1$:

In [6]:
sum(results)

[1] 1

Создадим вспомогательную функцию для взаимодействия с результатами.

In [7]:
if (!require("hash")) {
    install.packages("hash")
}
library(hash)
probabilities <- hash(results)

Loading required package: hash

hash-2.2.6.2 provided by Decision Patterns




In [8]:
P <- function(P_index) {
    hash_index <- paste("P_", P_index, sep = "")

    try(
        if (!has.key(hash_index, probabilities)) {
            stop("No P value with index: ", P_index)
        }
    )

    return(probabilities[[hash_index]])
}

#### Среднее число заявок в системе
Для вычисления среднего числа заявок в системе просуммируем произведения вероятностей
на соответствующие этим вероятностям значения заявок в системе
(длина очереди + количество заявок на проверке антивируса + количество
заявок на обслуживании):
$$
L = 0 \cdot P_{4000}(t) + 1 \cdot P_{3010}(t) + (1 + 1) \cdot P_{2110}(t) +
1 \cdot P_{3001}(t) + (1 + 1) \cdot P_{2011}(t) + (1 + 1 + 1) \cdot P_{1111}(t) +
2 \cdot P_{2002}(t) + (1 + 2) \cdot P_{1012}(t) + (1 + 1 + 2) \cdot P_{0112}(t) +
3 \cdot P_{1003}(t) + (1 + 3) \cdot P_{0013}(t)
$$

Упростив, получаем:
$$
L = P_{3010}(t) + 2 \cdot P_{2110}(t) +
P_{3001}(t) + 2 \cdot P_{2011}(t) + 3 \cdot P_{1111}(t) +
2 \cdot P_{2002}(t) + 3 \cdot P_{1012}(t) + 4 \cdot P_{0112}(t) +
3 \cdot P_{1003}(t) + 4 \cdot P_{0013}(t)
$$

In [9]:
mean_number_of_requests <- P("3010") + 2 * P("2110") + P("3001") + 2 * P("2011") +
    3 * P("1111") + 2 * P("2002") + 3 * P("1012") + 4 * P("0112") + 3 * P("1003") +
    4 * P("0013")
mean_number_of_requests

[1] 2.415041

#### Абсолютная пропускная способность
Для определения абсолютной пропускной способности учтем, что
заявки-программы, в которых обнаружен вирус не получают полного обслуживания
(отклоняются), т.е. нас интересует среднее число заявок получивших полное
обслуживание в единицу времени. Для замкнутых систем абсолютная пропускная
способность выражается как
$$
\lambda'=P_{\text{зан}}\cdot \mu
$$

In [10]:
absolute_flow_capacity <- mu * (
    3 * (P("1003") + P("0013")) +
    2 * (P("2002") + P("1012") + P("0112")) +
    P("3001") + P("2011") + P("1111")
)
absolute_flow_capacity

[1] 0.1697884

#### Среднее время пребывания заявки в системе
$$
T =\frac{L}{\lambda'}
$$

In [11]:
T <- mean_number_of_requests / absolute_flow_capacity
T

[1] 14.22383

### Численно
Построим с помощью пакета simmer симуляцию системы.

In [12]:
if (!require("simmer")) {
    install.packages("simmer")
}
library(simmer)

if (!require("parallel")) {
    install.packages("parallel")
}
library(parallel)

Loading required package: simmer

Loading required package: parallel



Зададим траекторию отказа в случае полной очереди:

In [13]:
antivirus_server_path <- trajectory("antivrus server path") %>%
    seize("antivirus_server", amount = 1) %>%
    timeout(function() rexp(1, nu)) %>%
    release("antivirus_server", amount = 1) %>%
    leave(
          function() runif(1) < p,
          out = trajectory() %>% log_("Antivirus found", level = 3)
    ) %>%
    log_("No antivirus found", level = 3)

main_server <- trajectory("main server path") %>%
    seize("server", amount = 1) %>%
    timeout(function() rexp(1, mu)) %>%
    release("server", amount = 1)

In [14]:
SIMULATION_TIME <- 10000

envs <- mclapply(1:30, function(i) {
    env <- simmer("SuperDuperSim", log_level = 2)

    programmers_source <- trajectory("programmers source path") %>%
            set_source("program", function() {
                number_of_free_programmers <- get_global(env, "number_of_free_programmers")

                if (number_of_free_programmers <= 0) {
                    return(0.5)
                } else {
                    return(rexp(1, lambda * number_of_free_programmers))
                }
            })

    programmers_path <- trajectory("program's path") %>%
            join(antivirus_server_path) %>%
            set_global("number_of_free_programmers", function() {
                on_server <- get_server_count(env, resources = c('antivirus_server', 'server'))
                in_queue <- get_queue_count(env, resources = c('antivirus_server', 'server'))

                in_system <- sum(on_server) + sum(in_queue)

                number_of_free_programmers <- K - in_system

                return(number_of_free_programmers)
            }) %>%
            join(programmers_source) %>%
            join(main_server) %>%
            set_global("number_of_free_programmers", value = 1, mod = "+") %>%
            join(programmers_source)

    return(env %>%
        add_resource(
            "antivirus_server",
            capacity = N,
            queue_size = m1
        ) %>%
        add_resource(
            "server",
            capacity = M,
            queue_size = m2
        ) %>%
        add_generator(
            "program",
            programmers_path,
            function() rexp(1, K * lambda)
        ) %>%
        run(until = SIMULATION_TIME) %>%
        wrap()
    )
})

In [15]:
arrivals <- get_mon_arrivals(envs)
arrivals

name        start_time end_time  activity_time finished replication
1      program0     1.994438   2.163248  0.1688102     TRUE    1          
2      program3     5.241048   5.520207  0.2791584     TRUE    1          
3      program6    11.698871  11.698871  0.0000000    FALSE    1          
4      program4    10.511292  14.282738  3.7714468     TRUE    1          
5      program5    11.252212  14.772749  1.7565678     TRUE    1          
6      program7    14.248803  15.804165  1.0314161     TRUE    1          
7      program10   18.161040  18.161040  0.0000000    FALSE    1          
8      program11   18.352782  18.352782  0.0000000    FALSE    1          
9      program13   19.455590  19.455590  0.0000000    FALSE    1          
10     program9    16.958711  20.013519  0.9570154    FALSE    1          
11     program15   23.157791  23.157791  0.0000000    FALSE    1          
12     program1     2.844267  23.240171 20.3959045     TRUE    1          
13     program12   19.360492  23.644481  3.6309614     TRUE    1          
14     program14   22.657791  23.964816  0.3203348     TRUE    1          
15     program17   25.155811  26.149472  0.3577639    FALSE    1          
16     program20   29.723845  29.723845  0.0000000    FALSE    1          
17     program18   28.615130  29.757789  1.1426589     TRUE    1          
18     program19   29.586527  31.008537  1.2507476     TRUE    1          
19     program2     3.937140  31.766366 27.8292265     TRUE    1          
20     program22   33.937974  35.310702  0.0417279    FALSE    1          
21     program16   23.657791  35.433498 11.4686820     TRUE    1          
22     program23   36.415379  38.068272  1.6528928     TRUE    1          
23     program21   32.593673  38.903567  6.3098948     TRUE    1          
24     program24   38.223266  39.961933  1.7386667     TRUE    1          
25     program28   46.012089  46.012089  0.0000000    FALSE    1          
26     program26   45.605848  46.018668  0.4128203     TRUE    1          
27     program29   46.475603  47.187946  0.2140243     TRUE    1          
28     program30   49.249649  50.734565  1.4849152    FALSE    1          
29     program31   50.592160  51.159336  0.4247709     TRUE    1          
30     program27   45.636112  56.227579 10.2089108     TRUE    1          
⋮      ⋮           ⋮          ⋮         ⋮             ⋮        ⋮          
222757 program7241 9964.888   9964.888   0.00000000   FALSE    30         
222758 program7242 9965.127   9965.127   0.00000000   FALSE    30         
222759 program7243 9965.575   9965.575   0.00000000   FALSE    30         
222760 program7244 9966.426   9966.426   0.00000000   FALSE    30         
222761 program7245 9966.836   9966.836   0.00000000   FALSE    30         
222762 program7247 9969.557   9969.557   0.00000000   FALSE    30         
222763 program7236 9961.994   9970.287   6.33879791    TRUE    30         
222764 program7249 9972.959   9972.959   0.00000000   FALSE    30         
222765 program7246 9968.816   9973.454   2.86920945    TRUE    30         
222766 program7248 9970.858   9974.487   1.03355247    TRUE    30         
222767 program7234 9961.739   9975.188  13.44903602    TRUE    30         
222768 program7253 9978.206   9978.206   0.00000000   FALSE    30         
222769 program7251 9976.848   9978.295   1.44742159    TRUE    30         
222770 program7255 9978.487   9978.487   0.00000000   FALSE    30         
222771 program7256 9979.439   9979.439   0.00000000   FALSE    30         
222772 program7250 9974.610   9979.450   4.83985970    TRUE    30         
222773 program7257 9979.737   9979.737   0.00000000   FALSE    30         
222774 program7258 9980.123   9980.123   0.00000000   FALSE    30         
222775 program7259 9980.501   9980.501   0.00000000   FALSE    30         
222776 program7260 9981.442   9981.442   0.00000000   FALSE    30         
222777 program7252 9978.164   9981.520   3.22429674    TRUE    30         
222778 program7262 9984.118   99

Логи симуляции:

In [16]:
resources <- get_mon_resources(envs)
resources

resource         time      server queue capacity queue_size system limit
1      antivirus_server  1.994438 1      0     1        1          1      2    
2      antivirus_server  2.163248 0      0     1        1          0      2    
3      antivirus_server  2.844267 1      0     1        1          1      2    
4      antivirus_server  3.392928 0      0     1        1          0      2    
5      server            3.392928 1      0     3        0          1      3    
6      antivirus_server  3.937140 1      0     1        1          1      2    
7      antivirus_server  5.066169 0      0     1        1          0      2    
8      server            5.066169 2      0     3        0          2      3    
9      antivirus_server  5.241048 1      0     1        1          1      2    
10     antivirus_server  5.520207 0      0     1        1          0      2    
11     antivirus_server 10.511292 1      0     1        1          1      2    
12     antivirus_server 11.252212 1      1     1        1          2      2    
13     antivirus_server 13.016181 1      0     1        1          1      2    
14     server           13.016181 3      0     3        0          3      3    
15     antivirus_server 14.248803 1      1     1        1          2      2    
16     server           14.282738 2      0     3        0          2      3    
17     antivirus_server 14.772749 1      0     1        1          1      2    
18     antivirus_server 15.804165 0      0     1        1          0      2    
19     antivirus_server 16.522822 1      0     1        1          1      2    
20     antivirus_server 16.958711 1      1     1        1          2      2    
21     antivirus_server 19.056504 1      0     1        1          1      2    
22     server           19.056504 3      0     3        0          3      3    
23     antivirus_server 19.360492 1      1     1        1          2      2    
24     antivirus_server 20.013519 1      0     1        1          1      2    
25     antivirus_server 22.657791 1      1     1        1          2      2    
26     server           23.240171 2      0     3        0          2      3    
27     antivirus_server 23.644481 1      0     1        1          1      2    
28     antivirus_server 23.657791 1      1     1        1          2      2    
29     antivirus_server 23.964816 1      0     1        1          1      2    
30     antivirus_server 25.155811 1      1     1        1          2      2    
⋮      ⋮                ⋮         ⋮      ⋮     ⋮        ⋮          ⋮      ⋮    
375343 antivirus_server 9978.164  1      1     1        1          2      2    
375344 antivirus_server 9978.295  1      0     1        1          1      2    
375345 antivirus_server 9978.338  1      1     1        1          2      2    
375346 server           9979.450  1      0     3        0          1      3    
375347 antivirus_server 9981.520  1      0     1        1          1      2    
375348 antivirus_server 9982.771  1      1     1        1          2      2    
375349 antivirus_server 9985.172  1      0     1        1          1      2    
375350 server           9985.172  2      0     3        0          2      3    
375351 antivirus_server 9985.245  1      1     1        1          2      2    
375352 antivirus_server 9985.283  1      0     1        1          1      2    
375353 antivirus_server 9985.758  0      0     1        1          0      2    
375354 server           9985.758  3      0     3        0          3      3    
375355 antivirus_server 9987.667  1      0     1        1          1      2    
375356 antivirus_server 9987.732  0      0     1        1          0      2    
375357 server           9989.301  2      0     3        0          2      3    
375358 antivirus_server 9990.647  1      0     1        1          1      2    
375359 antivirus_server 9990.650  1      1     1        1          2      2    
375360 server           9990.769  1      0     3        0          1      3    
375361 

#### Абсолютную пропускную способность
Абсолютную пропускную способность найдем следующим образом:

In [17]:
number_of_finished <- arrivals %>% with(sum(finished))
number_of_finished

[1] 123825

In [18]:
finish_probability <- number_of_finished / nrow(arrivals)
finish_probability

[1] 0.5558024

In [19]:
absolute_flow_capacity <- lambda * finish_probability
absolute_flow_capacity

[1] 0.1580699

#### Среднее число заявок в системе
Рассчитаем значение, воспользовавшись таблицей использования ресурсов:

In [20]:
mean_number_of_requests <- resources %>% with(
    mean(queue) + mean(server) + mean(system)
)
mean_number_of_requests

[1] 2.437417

#### Среднее время нахождения заявок в системе
Найдем на основании практически величин среднее время нахождения заявок
в системе

In [21]:
T <- mean_number_of_requests / absolute_flow_capacity
T

[1] 15.41987

## Вывод
Как видно, теоретически вычисленное значение с некоторой точностью совпадает
со значением, полученным теоретически. При увеличении количества
экспериментов 𝑁 точность только увеличивается.

### Теоретически методом укрупненных состояний
Проанализируем сложную систему по частям, с помощью укрупненных (агрегированных) моделей.
В каждой модели подробно представим только некоторую часть системы, а влияние
остальных частей отразим некоторым обобщенным параметром (параметром связи).

Предложим декомпозицию нашей задачи в виде двух моделей.

В первой модели совокупность серверов и компьютеров по проверке наличия
вирусов заменим только одним обобщенный параметром - интенсивностью
обслуживания общ $\mu_{\text{общ}}$.

![Programmers graph](./Programmers_graph.png)

Как видно, модель будет обычной моделью
рождения гибели для замкнутой системы

$$
P_0 = \left(1 + \frac{K\lambda}{\mu_{\text{общ}}} + \frac{K(K-1)\lambda^2}{\mu_{\text{общ}}^2} + \ldots + \frac{K!\lambda^K}{\mu_{\text{общ}}^K}\right)^{-1}
$$

#### Итерация 1
Пусть $\mu_{\text{общ}} = 0.15 < \lambda'$. Тогда:

In [67]:
mu_ <- 0.15
y_ <- lambda / mu_
y_

[1] 1.895996

Каждый член суммы можно также представить:
$$
\text{sum_part}_i =\frac{K!}{(K-i)!} \cdot \frac{\lambda}{\mu_{\text{общ}}}
$$

In [68]:
i <- 0:K
sum_parts <- c(factorial(K) / factorial(K - i) * y_^i)
sum_parts

[1]   1.000000   7.583986  43.137627 163.577571 310.142482

In [69]:
P0 <- (sum(sum_parts)
)^(-1)
P0

[1] 0.001903161

$$
P_1 = P_0 \frac{K\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_1 \\
P_2 = P_1 \frac{(K - 1)\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_2 \\
P_3 = P_2\frac{(K - 2)\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_3 \\
P_i = P_{i-1}\frac{(K - (i - 1))\lambda}{\mu_{\text{общ}}} = P_0 \cdot
\text{sum_part}_i
$$

In [71]:
Pi <- unlist(lapply(sum_parts, function(sum_part) P0 * sum_part))
Pi

[1] 0.001903161 0.014433544 0.082097843 0.311314426 0.590251026

Проверим, что вероятности имеют адекватные значения.

In [72]:
sum(Pi)

[1] 1

$$
L_{\text{сист}} = \sum_{k=1}^K k \cdot P_k
$$

In [73]:
mean_number_of_requests1 <- sum(c(1:K) * Pi[2:(K+1)])
mean_number_of_requests1

[1] 3.473577

Во второй модели будем считать, что в системе постоянно циркулируют
$L_{\text{сист}}$ заявок. В качестве состояний системы возьмем количество
заявок, обрабатываемых на основных серверах:
$$
S_0, S_1, S_2, \ldots, S_{L_{\text{сист}}}
$$

![Servers graph](./Servers_graph.png)

В данной итерации, округляя, мы получили $[L_{\text{сист}} \approx 3.47] = 3$

$$
\nu_n = (1-p)\nu \cdot \min{(N,L_{\text{сист}}-n)}, n=0, \ldots, L_{\text{сист}} - 1 \\
$$

In [74]:
mean_number_of_requests1 <- round(mean_number_of_requests1)
n <- 0:(mean_number_of_requests1 - 1)

nu_n <- unlist(lapply(n, function(n) (1 - p) * nu * min(N, mean_number_of_requests1 - n)))
nu_n

[1] 0.3309272 0.3309272 0.3309272

$$
\mu_k = \mu \cdot \min{(M, k)}, k=1, 2, \ldots, L_{\text{сист}}
$$

In [75]:
k <- 1:mean_number_of_requests1

mu_k <- unlist(lapply(k, function(k) mu * min(M, k)))
mu_k

[1] 0.1046501 0.2093003 0.3139504

Аналогичным образом посчитаем финальные вероятности:
$$
\pi_0 = \left(1+\frac{\nu_0}{\mu_1}+\frac{\nu_0\nu_1}{\mu_1\mu_2}+\ldots\right)^{-1} \\
\pi_1=\pi_0\cdot \frac{\nu_0}{\mu_1} \\
\pi_2=\pi_0\cdot \frac{\nu_0\nu_1}{\mu_1\mu_2} \\
\dots
$$

In [76]:
sum_parts <- c()
sum_parts[1] <- 1
sum_parts[2] <- sum_parts[1] * nu_n[1] / mu_k[1]
sum_parts[3] <- sum_parts[2] * nu_n[2] / mu_k[2]
sum_parts[4] <- sum_parts[3] * nu_n[3] / mu_k[3]

pi_0 <- (sum(sum_parts))^(-1)
pi_0

[1] 0.06928926

In [77]:
pi_i <- unlist(lapply(sum_parts, function(sum_part) pi_0 * sum_part))
pi_i

[1] 0.06928926 0.21910818 0.34643460 0.36516796

Проверим, что вероятности имеют адекватные значения.

In [78]:
sum(pi_i)

[1] 1

Тогда:
$$
\mu_{\text{общ}} = \sum_{n=1}^{L_{\text{сист}}}\pi_n\cdot \mu_n
$$

In [79]:
mu_ <- sum(mu_k[1:(mean_number_of_requests1)] * pi_i[2:(mean_number_of_requests1 + 1)])
mu_

[1] 0.2100832

#### Итерация 2

In [80]:
y_ <- lambda / mu_
y_

[1] 1.353747

Каждый член суммы можно также представить:
$$
\text{sum_part}_i =\frac{K!}{(K-i)!} \cdot \frac{\lambda}{\mu_{\text{общ}}}
$$

In [81]:
i <- 0:K
sum_parts <- c(factorial(K) / factorial(K - i) * y_^i)
sum_parts

[1]  1.000000  5.414988 21.991570 59.542044 80.604861

In [82]:
P0 <- (sum(sum_parts)
)^(-1)
P0

[1] 0.005932836

$$
P_1 = P_0 \frac{K\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_1 \\
P_2 = P_1 \frac{(K - 1)\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_2 \\
P_3 = P_2\frac{(K - 2)\lambda}{\mu_{\text{общ}}} = P_0 \cdot \text{sum_part}_3 \\
P_i = P_{i-1}\frac{(K - (i - 1))\lambda}{\mu_{\text{общ}}} = P_0 \cdot
\text{sum_part}_i
$$

In [83]:
Pi <- unlist(lapply(sum_parts, function(sum_part) P0 * sum_part))
Pi

[1] 0.005932836 0.032126233 0.130472373 0.353253161 0.478215397

Проверим, что вероятности имеют адекватные значения.

In [84]:
sum(Pi)

[1] 1

$$
L_{\text{сист}} = \sum_{k=1}^K k \cdot P_k
$$

In [110]:
mean_number_of_requests1 <- sum(c(1:K) * Pi[2:(K+1)])
mean_number_of_requests1

[1] 3.472403

В данной итерации, округляя, мы получили $[L_{\text{сист}} \approx 3.47]
= 3$. Ясно, что снова выбрав 3, ничего не поменяется, поэтому попробуем
$L_{\text{сист}} = 2$.

$$
\nu_n = (1-p)\nu \cdot \min{(N,L_{\text{сист}}-n)}, n=0, \ldots, L_{\text{сист}} - 1 \\
$$

In [128]:
mean_number_of_requests1 <- 2
n <- 0:(mean_number_of_requests1 - 1)

nu_n <- unlist(lapply(n, function(n) (1 - p) * nu * min(N, mean_number_of_requests1 - n)))
nu_n

[1] 0.3309272 0.3309272

$$
\mu_k = \mu \cdot \min{(M, k)}, k=1, 2, \ldots, L_{\text{сист}}
$$

In [88]:
k <- 1:mean_number_of_requests1

mu_k <- unlist(lapply(k, function(k) mu * min(M, k)))
mu_k

[1] 0.1046501 0.2093003

Аналогичным образом посчитаем финальные вероятности:
$$
\pi_0 = \left(1+\frac{\nu_0}{\mu_1}+\frac{\nu_0\nu_1}{\mu_1\mu_2}+\ldots\right)^{-1} \\
\pi_1=\pi_0\cdot \frac{\nu_0}{\mu_1} \\
\pi_2=\pi_0\cdot \frac{\nu_0\nu_1}{\mu_1\mu_2} \\
\dots
$$

In [89]:
sum_parts <- c()
sum_parts[1] <- 1
sum_parts[2] <- sum_parts[1] * nu_n[1] / mu_k[1]
sum_parts[3] <- sum_parts[2] * nu_n[2] / mu_k[2]

pi_0 <- (sum(sum_parts))^(-1)
pi_0

[1] 0.1091458

In [90]:
pi_i <- unlist(lapply(sum_parts, function(sum_part) pi_0 * sum_part))
pi_i

[1] 0.1091458 0.3451435 0.5457106

Проверим, что вероятности имеют адекватные значения.

In [91]:
sum(pi_i)

[1] 1

Тогда:
$$
\mu_{\text{общ}} = \sum_{n=1}^{L_{\text{сист}}}\pi_n\cdot \mu_n
$$

In [92]:
mu_ <- sum(mu_k[1:(mean_number_of_requests1)] * pi_i[2:(mean_number_of_requests1 + 1)])
mu_

[1] 0.1503367

$$
T_{\text{сист}} = \frac{L_{\text{сист}}}{\lambda(K-L_{\text{сист}})}
$$

In [123]:
T <- mean_number_of_requests1 / (lambda * (K - mean_number_of_requests1))
T

[1] 14.06472

#### Итого

In [130]:
c(mu_, T)

[1]  0.1503367 14.0647244

### Выводы
Как видно, метод укрупнённых состояний позволяет упростить вычисления в очень
больших системах. Однако приходится пожертвовать точностью вычислений, что
связано с округлением $L_{\text{сист}}$ до целых.